### PR TITLE
An early implementation of basefunctions for databasemanagement. 

### PR DESCRIPTION
**Pull request by MattKa01**
Function for setting up a connection, searching for documents and deleting documents. Most of the documentvalues are hardcoded but can be changed if you want to use the code for testing. Also see the README for instructions on usage.

### CODE REVIEW

**Review by MekkerMax — DISMISSED** [review]

File: lib/db_function_enum.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package lib
+
+type DBFunction int
+
+const (
+	FnFindOne DBFunction = iota + 1 //EnumIndex = 1
```
Comment (Jacalz) [inline lib/db_function_enum.go]: If you don't have a need for it to start on 1, start on 0.

File: lib/db_function_enum.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+	FnDeleteOne //EnumIndex = 3
+
+	FnDeleteMany //EnumIndex = 4
+
+	FnInsertOne //EnumIndex = 5
+
+	FnInsertMany //EnumIndex = 6
+
+	FnFindAll //EnumIndex = 7
```
Comment (Jacalz) [inline lib/db_function_enum.go]: You can remove the comments with index. You should not hard-code the numbers to the index is irrelevant.

File: lib/db_function_enum.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
+package lib
+
+type DBFunction int
```
Comment (Jacalz) [inline lib/db_function_enum.go]: Public types should have a doc comment.

File: lib/db_update.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+// This function ignores the namspace not found error so it won't crash
+// the program if the database doesn't exist.
+// Keep in mind that this function is irreversible so be careful when using.
+func DropDatabase(client mongo.Client, database string) {
+
+	err := client.Database(database).Drop(context.TODO())
+	if err != nil {
+		panic(err)
+	}
+}
```
Comment (Jacalz) [inline lib/db_update.go]:
```suggestion
func DropDatabase(client mongo.Client, database string) error {
        return client.Database(database).Drop(context.TODO())
}
```

File: lib/db_function_enum.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+		"FnDeleteMany",
+		"FnInsertOne",
+		"FnInsertMany",
+		"FnFindAll"}[f-1]
+}
+
+// EnumIndex - Creating common behavior - give the type a EnumIndex function
+func (f DBFunction) EnumIndex() int {
+	return int(f)
+}
```
Comment (Jacalz) [inline lib/db_function_enum.go]: Why? The type is an int already.

File: lib/db_fetch.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+	"encoding/json"
+	"fmt"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+)
+
+func FetchHello() {
+	fmt.Println("Hello, this is the fetch package")
+}
```
Comment (Jacalz) [inline lib/db_fetch.go]: Remove test print please

**Review by Jacalz — CHANGES_REQUESTED** [review]
I added a few comments inline. I would also appreciate if the "lib" package gets renamed to something better describing. It would also be nice to avoid panics and instead have functions return error values.

File: main.go [hunk 7]
```diff
@@ -1,7 +1,83 @@
...
+	//document := bson.D{
+	//	{"Time", "14:52:22"},
+	//	{"Roll", 0.723491},
+	//	{"Pitch", -3.248201},
+	//	{"Yaw", 0.345234}}
+
+	//Example of a filter
+	//filter := bson.D{{"Time", "15:40:22"}}
+
+	lib.TestConnection(*client)
```
Comment (Jacalz) [inline main.go]: Doing `*client` grabs the value and copies it. Should the function take a `*mongo.Client` instead.

**Review by Jacalz — CHANGES_REQUESTED** [review]

File: lib/db_fetch.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+// TODO: Find a way to decode hexadecimal in error messages
+// FetchDocument querys the database and returns matching documents.
+// An open clientconnection is required in the global scope to fetch from the database.
+// Both the database and collection name must be given, as well as a query filter.
+// The fetch mode can be set by choosing the function enum
+// It will return the result as marshaled jsondata
+func FetchDocument(client mongo.Client, database string, collection string, filter bson.D, fn DBFunction) []byte {
```
Comment (Jacalz) [inline lib/db_fetch.go]: `fn DBFunction` can probably be `many bool`.

File: lib/db_info.go [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+	fmt.Println("Hello, this is the connect package")
+}
+
+// TestConnection sends a ping to the database via the client instance
+// It panics if the connection is down, otherwise prints a confirmation
+func TestConnection(client mongo.Client) {
+
+	// Send a ping to confirm a successful connection
+	var result bson.M
+	if err := client.Database("admin").RunCommand(context.TODO(), bson.D{{"ping", 1}}).Decode(&result); err != nil {
```
Comment (Jacalz) [inline lib/db_info.go]: TODO: We might want to use a context with timeout.

File: lib/db_update.go [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+	fmt.Println("Hello, this is the update package")
+}
+
+// TODO: find a better solution to shifting between bson.D and []interface{}
+// InsertDocument takes the client, a collection name and a bson document to be inserted.
+// The client must have an open connection in the global scope for this to work.
+// The function will print the unique id of the inserted document
+// Keep in mind that the document must be converted to a bson D type before calling
+// this function.
+func InsertDocument(client mongo.Client, database string, collection string, document bson.D, docs []interface{}, fn DBFunction) {
```
Comment (Jacalz) [inline lib/db_update.go]:
```suggestion
func InsertDocument(client mongo.Client, database string, collection string, document bson.D, docs []any, fn DBFunction) {
```

**Review by Jacalz — CHANGES_REQUESTED** [review]

**Comment by MattKa01** [comment]
All the comments have pretty much been handled. Currently the client object is used the same way since I need to read more into it before making changes. The program will also still panic upon a error which is good for testing but will be changed to slog after reading the documentation.

The enum has been removed and seperate functions for the one/many calls have been implemented.

The lib package has been renamed to db

File: db/db_update.go [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+// DeleteManyDocuments deletes all the documents that matches the filter given
+// The function also takes a client, the name of the collection, the databasename
+// Keep in mind that the client most have an open connection in the global scope for this
+// to work.
+// To delete all documents it can be called with an empty bson.D object filter although it would be
+// better to drop the collection
+// The function will print the amount of documents deleted
+func DeleteManyDocuments(client mongo.Client, database string, collection string, filter bson.D) {
```
Comment (Jacalz) [inline db/db_update.go]: I am pretty sure that all of these functions should take a `client *mongo.Client` as it is what is returned from `mongo.Connect`. All of the `*client` dereferences when calling the functions makes me very nervous.

**Review by Jacalz — CHANGES_REQUESTED** [review]
Thanks. This is looking a lot better. I have one comment inline that wasn't fixed from my older set of review comments.

**Comment by Jacalz** [comment]
You need to rebase on main and fix the merge conflicts. Once the workflows are passing and the comment above is resolved, this should be ready to be merged.

**Comment by Jacalz** [comment]
go install mvdan.cc/gofumpt@latest

gofumpt -w .

**Review by Jacalz — APPROVED** [review]

**Pull request merged**